### PR TITLE
[New metric] Index Redundancy Metric

### LINKED ISSUE #79: New Metric - Index Redundancy Metric
Metric name
----------------
Index Redundancy Metric [Storage]

Metric description
--------------------
The index Redundancy Metric will be responsible for finding the unused, duplicate, similar and overlapping indexes in a database. This will also help us to identify unused tables. This will result in quicker space savings and better CPU Performance.

Additional details for the metric
---------------------------------

### PR DESCRIPTION
**Pull request by prajain12**
## Motivation and Context

The index Redundancy Metric will be responsible for finding the unused, duplicate, similar and overlapping indexes in a database. This will also help us to identify unused tables. This will result in quicker space savings and better CPU Performance.

## Description

Index Redundancy Metric and dependencies added

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Backward compatibility

- [x] snorql-extensions
- [ ] snorql-framework

## How Has This Been Tested?

Tested locally using udaan sql-monitoring

## Migration and deployment details (if appropriate)

Not required.

### CODE REVIEW

File: snorql-extensions/src/main/kotlin/com/udaan/snorql/extensions/storage/metrics/DbIndexRedundancyMetric.kt [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        if (shorterString == null) return true
+        val longerArr = longerString.split(" ")
+        val shortArr = shorterString.split(" ")
+        if (longerArr.size < shortArr.size) return false
+        shortArr.forEachIndexed { index, element ->
+            if (element != longerArr[index]) {
+                return false
+            }
+        }
+        return true
```
Comment (lokeshdevnani) [inline snorql-extensions/src/main/kotlin/com/udaan/snorql/extensions/storage/metrics/DbIndexRedundancyMetric.kt:235]: This seems too complicated. Please use kotlin methods.

File: snorql-extensions/src/main/kotlin/com/udaan/snorql/extensions/storage/metrics/DbIndexRedundancyMetric.kt [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+        )
+        if (checkStartsWithWord(
+                parentIndex.indexColumnNrs,
+                childIndex.indexColumnNrs
+            ) // && (parentIncludeCols subtract childIncludeCols).isEmpty()
+            && (childIncludeCols subtract parentIncludeCols).isEmpty()
+        ) {
+            return true
+        }
+        return false
```
Comment (lokeshdevnani) [inline snorql-extensions/src/main/kotlin/com/udaan/snorql/extensions/storage/metrics/DbIndexRedundancyMetric.kt:198]:
```suggestion
        return (checkStartsWithWord(
                parentIndex.indexColumnNrs,
                childIndex.indexColumnNrs
            ) && (childIncludeCols subtract parentIncludeCols).isEmpty()) 
```

File: snorql-extensions/src/main/kotlin/com/udaan/snorql/extensions/storage/metrics/DbIndexRedundancyMetric.kt [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+    }
+
+    private fun isSimilar(parentIndex: DbIndexRedundancyDTO, childIndex: DbIndexRedundancyDTO): Boolean {
+        if (parentIndex.indexColumnNrs == childIndex.indexColumnNrs
+            && parentIndex.includeColumnNrs != childIndex.includeColumnNrs
+        ) {
+            return true
+        }
+        return false
```
Comment (lokeshdevnani) [inline snorql-extensions/src/main/kotlin/com/udaan/snorql/extensions/storage/metrics/DbIndexRedundancyMetric.kt:207]: `if (bool_condition) { return true } else { return false }` is a redundant expression.
Always use `return bool_condition`

**Review by lokeshdevnani — CHANGES_REQUESTED** [review]
Left a few comments. 
Please add tests for the index redundancy compute logic.

File: snorql-extensions/src/main/kotlin/com/udaan/snorql/extensions/storage/metrics/DbIndexRedundancyMetric.kt (hunk 3 above)
Comment (prajain12) [reply snorql-extensions/src/main/kotlin/com/udaan/snorql/extensions/storage/metrics/DbIndexRedundancyMetric.kt]: Resolved

File: snorql-extensions/src/main/kotlin/com/udaan/snorql/extensions/storage/metrics/DbIndexRedundancyMetric.kt (hunk 1 above)
Comment (prajain12) [reply snorql-extensions/src/main/kotlin/com/udaan/snorql/extensions/storage/metrics/DbIndexRedundancyMetric.kt]: Can you please suggest optimisations on this?

**Review by lakshmi-udaan — APPROVED** [review]

**Review by lokeshdevnani — APPROVED** [review]

**Pull request merged**